Implement non-standard pthread_tryjoin and pthread_timedjoin.

ClaCodes opened the pull request:
Implement the non-standard facilities for joining pthreads.

## Open Questions
- [x] naming `_np` for non-standard as gnu does?
- [x] test allowed to sleep 300 ms?
- [x] use maximum value of `time_t` to represent `forever` in the timespec, which is the correct/portable makro?
- [x] convert `timespec` to `timeout` -> handle overflow?

This is my first PR, so let me know, if something formal is wrong. @cfriedt please let me know your opinion on this.

ClaCodes committed:
posix: pthread: implement non-standard try-join and timed-join

These functions can be used to join pthreads in a non-standard way.
The function pthread_tryjoin will not block and simply test whether the
thread has exited already. The function pthread_timed_join will only block
until the specified time. The functions are wrappers for calling the
k_thread_join with timeout K_NO_WAIT and with a specific timeout as opposed
to calling it with K_FOREVER.

Signed-off-by: Cla Galliard <clamattia@gmail.com>

cfriedt commented:
@ClaCodes - I would definitely add the `_np` suffix.

I'll add a more detailed review in a few minutes.

cfriedt commented on include/zephyr/posix/pthread.h:
```diff
@@ -421,6 +421,8 @@ int pthread_attr_setinheritsched(pthread_attr_t *attr, int inheritsched);
 int pthread_once(pthread_once_t *once, void (*initFunc)(void));
 #endif
 FUNC_NORETURN void pthread_exit(void *retval);
+int pthread_timedjoin_np(pthread_t thread, void **status, const struct timespec abstime);
```
```suggestion
int pthread_timedjoin_np(pthread_t thread, void **status, const struct timespec *abstime);
```

cfriedt commented on lib/posix/options/pthread.c:
```diff
@@ -1046,20 +1046,34 @@ void pthread_exit(void *retval)
+ * Non-portable GNU extension of IEEE 1003.1
  */
-int pthread_join(pthread_t pthread, void **status)
+int pthread_timedjoin_np(pthread_t pthread, void **status, const struct timespec *abstime)
```
One possibly cleaner way to go about this is to convert all absolute `struct timespec` to `k_timeout` at the point of entry, and have an internal `pthread_timedjoin_internal(pthread_t th, void **status, k_timeout_t timeout)` (where the timeout is already specified as a `k_timeout_t` instead of `struct timespec`).

That way, e.g.
* `pthread_join(th, st)` -> `pthread_timedjoin_internal(th, st, K_FOREVER)`
* `pthread_tryjoin(th, st)` -> `pthread_timedjoin_internal(th, st, K_NO_WAIT)`
* `pthread_timedjoin_np(th, st, to)` -> `pthread_timedjoin_internal(th, st, K_MSEC(timespec_to_timeoutms(to)))`

cfriedt reviewed (changes requested):
I think we should convert to `k_timeout` at the point of entry, rename `pthread_join()` to static `pthread_timedjoin_internal()`,  also accept a `k_timeout_t` timeout, and then the other join functions can be 1-line wrappers for the internal version.

Additionally, please add tests in a second commit and follow the commit guidelines.

Tests should be directed to individual functions, so add

`ZTEST(pthread, test_pthread_timedjoin)`
`ZTEST(pthread, test_pthread_tryjoin)`

following the existing conventions (see git log), where each test case tests various permutations of invalid input and then valid input.

cfriedt commented:
@ClaCodes - be sure to tests prior to pushing via

```shell
twister -i -T tests/posix
```
and check compliance via

```shell
./scripts/checkpatch.pl -g <commit-before-yours>..
```

ClaCodes committed:
tests/posix: Add test for pthread_timed_join and pthread_try_join

Test for the non-standard but useful pthread_timed_join- and
pthread_try_join-functions.

Signed-off-by: Cla Galliard <clamattia@gmail.com>

ClaCodes commented:
@cfriedt I tried to address your suggestions and used `checkpatch`. I am unable to run the test locally because I am on an unsupported system. Will have to rely on the CI for that currently.

cfriedt commented on lib/posix/options/pthread.c:
```diff
@@ -1098,9 +1093,7 @@ int pthread_join(pthread_t pthread, void **status)
-	ret = k_thread_join(&t->thread, K_FOREVER);
-	/* other possibilities? */
-	__ASSERT_NO_MSG(ret == 0);
+	ret = k_thread_join(&t->thread, timeout);
```
We need to process the return value of `k_thread_join()`.

This should work - what do you think @ycsin ?

```suggestion
	ret = k_thread_join(&t->thread, timeout);
	if (ret < 0) {
		return ETIMEDOUT;
  }
```

ClaCodes replied:
Two questions:
- I understand your suggestion, that we need not `posix_thread_recycle();` in case we could not join?
- I took this as reference: https://www.man7.org/linux/man-pages/man3/pthread_timedjoin_np.3.html  In that case we should return `-EBUSY` if we did not wait at all (`try_join` for example) and `-ETIMEDOUT` if we timed out. This is almost what https://docs.zephyrproject.org/latest/kernel/services/threads/index.html#c.k_thread_join does. Except it returns `-EAGAIN`. Should we return `-EAGAIN` as `k_thread_join` does or `-ETIMEDOUT` as the reference suggests.

```suggestion
	ret = k_thread_join(&t->thread, timeout);
	if (ret < 0) {
		return ret;
        }
```
```suggestion
	ret = k_thread_join(&t->thread, timeout);
	if (ret == -EAGAIN) {
		return -ETIMEDOUT;
        } else if (ret < 0) {
		return ret;
	}        
```

ClaCodes replied:
For proposal see latest update. I brought the assertion back for `-EDEADLK`.

cfriedt reviewed (changes requested):
I think we need to check the return value of `k_thread_join()` now that the assertion is gone.

Otherwise, looks ok.

I'm just on mobile atm, so cannot look in detail.

Please double check that all meaningful input permutations are made in the testsuite as well.

ClaCodes commented:
> Please double check that all meaningful input permutations are made in the testsuite as well.

There is probably more that could be done. For now idea was:
- tryjoin:
  - spawn thread that ends after 1s
  - immediately `tryjoin` expect `EBUSY`
  - sleep 1s
  - `tryjoin` expect `success`

- timedjoin:
  - create thread that ends after 1s
  - Test `-EINVALID` for negativ seconds, negative nanoseconds and `nanonseconds == NSEC_PER_SEC`
  - timed join with timeout less then 1s after start, excpect `-ETIMEDOUT`
  - timed join with timeout later then 1s after start, excpect success

cfriedt commented on lib/posix/options/pthread.c:
```diff
@@ -1114,6 +1115,41 @@ int pthread_join(pthread_t pthread, void **status)
+ */
+int pthread_timedjoin_np(pthread_t pthread, void **status, const struct timespec *abstime)
+{
+	if (abstime->tv_sec < 0 || abstime->tv_nsec < 0 || abstime->tv_nsec >= NSEC_PER_SEC) {
```
This should have test coverage as well
 ```suggestion
	if (abstime == NULL) {
		return EINVAL;
	}
	if (abstime->tv_sec < 0 || abstime->tv_nsec < 0 || abstime->tv_nsec >= NSEC_PER_SEC) {
```

cfriedt reviewed (commented):
A null check would be good before dereferencing abstime

cfriedt commented on tests/posix/common/src/pthread.c:
```diff
@@ -371,6 +378,66 @@ ZTEST(pthread, test_pthread_termination)
+	sleep(ONE_SECOND);
+
+	/* Attempting to join immediately should succeed now */
+	zassert_ok(pthread_tryjoin_np(alive_1_s, &retval));
```
Looks like this is not succeeding in tests

cfriedt commented on tests/posix/common/src/pthread.c:
```diff
@@ -371,6 +378,66 @@ ZTEST(pthread, test_pthread_termination)
+	zassert_equal(pthread_timedjoin_np(alive_1_s, &ret, &not_done), ETIMEDOUT);
+
+	/* Attempting to join with a timeout, when the thread is done, should succeed */
+	zassert_ok(pthread_timedjoin_np(alive_1_s, &ret, &done));
```
This test looks like it's failing.

cfriedt replied:
I'll check in on this again later.

cfriedt commented on lib/posix/options/pthread.c:
```diff
@@ -1114,6 +1115,44 @@ int pthread_join(pthread_t pthread, void **status)
+		return EINVAL;
+	}
+
+	return pthread_timedjoin_internal(pthread, status, K_MSEC(timespec_to_timeoutms(abstime)));
```
It looks like this absolute timespec should be converted to a relative delay. Before the call to timedjoin_internal, maybe subtract `k_uptime_get()` from the milliseconds value?

ClaCodes replied:
I thought that was the purpose of `timespec_to_timeoutms`. It subtracts `current time` internally.

cfriedt replied:
Yes, sorry - my bad.

ClaCodes replied:
btw: This gives us essentially only msec resolution. Not sure, if a problem or not.

ycsin replied:
I guess it should be fine, whoever needs finer resolution can probably submit a patch later.

cfriedt reviewed (commented):
Probably could use a rebase, and there are a couple of platforms where tests are not passing.

cfriedt commented:
@ClaCodes - would be nice to get this in - can you make some changes and resubmit?

ClaCodes commented:
> @ClaCodes - would be nice to get this in - can you make some changes and resubmit?

I rebased and tried to make test more robust by adding a bit more time. I am not sure, why some test are failing. I might have some time to look into it more, if adding more time was not good enough.
Maybe you could have a look at the test to see if I have some wrong assumptions.

cfriedt commented:
@ClaCodes - I think it's really common to not know where to start debugging these kinds of bugs that show up in CI, but if there is one piece of advice that I may give, it would be to start debugging with one platform that is failing.

I'll see if I can point out a good one...

cfriedt commented:
@ClaCodes - can you please rebase? It would be great to get this merged.

ClaCodes commented:
I have rebased.
I need help with the failing tests. Probably tests are failing due to my wrong assumptions about how time passes in the unit-tests.

cfriedt commented on lib/posix/options/pthread.c:
```diff
@@ -1092,8 +1088,13 @@ int pthread_join(pthread_t pthread, void **status)
-	ret = k_thread_join(&t->thread, K_FOREVER);
-	/* other possibilities? */
+	ret = k_thread_join(&t->thread, timeout);
+	if (ret == -EBUSY) {
+		return EBUSY;
+	} else if (ret == -EAGAIN) {
+		return ETIMEDOUT;
+	}
+	/* Can only be ok or -EDEADLK, which should never occur for pthreads */
 	__ASSERT_NO_MSG(ret == 0);
```
```suggestion
	if (ret == -EAGAIN) {
		return ETIMEDOUT;
	} else if (ret < 0) {
		/* since this is a non-portable API, we can also return EBUSY or EDEADLK */
		return -ret;
	}
```

cfriedt commented on tests/posix/common/src/pthread.c:
```diff
@@ -371,6 +378,63 @@ ZTEST(pthread, test_pthread_termination)
+	zassert_equal(pthread_tryjoin_np(alive_1_s, &retval), EBUSY);
+
+	/* Sleep so alive_1_s-thread will exit */
+	sleep(ONE_SECOND);
+	sleep(ONE_SECOND);
```
This could work, I guess.
```suggestion
	usleep(200 * USEC_PER_MSEC);
```

cfriedt commented on tests/posix/common/src/pthread.c:
```diff
@@ -143,6 +143,13 @@ static void *thread_top_exec(void *p1)
 
+static void *thread_alive_1_s(void *p1)
+{
+	sleep(ONE_SECOND);
```
Rather than 1s, it might be feasible to sleep for 100ms.

The only reason I suggest this, is that long delays in CI tend to grow significantly when there are several testsuite configurations and permutations.

But the

cfriedt commented on tests/posix/common/src/pthread.c:
```diff
@@ -371,6 +378,63 @@ ZTEST(pthread, test_pthread_termination)
 
+ZTEST(pthread, test_pthread_tryjoin)
+{
+	pthread_t alive_1_s = {0};
```
It's probably fine to just use `pthread_t th`.

cfriedt commented:
@ClaCodes - this is pretty close - just needs a nudge past the finish line 🏁

ClaCodes commented:
Unfortunately, I am still having trouble with the tests. If someone wants to pick this up, you are welcome to. Otherwise I will be closing.

Thanks for understanding and sorry.

cfriedt commented on tests/posix/common/src/pthread.c:
```diff
@@ -371,6 +377,67 @@ ZTEST(pthread, test_pthread_termination)
+	/* setup timespecs when the thread is still running and when it is done */
+	clock_gettime(CLOCK_MONOTONIC, &not_done);
+	clock_gettime(CLOCK_MONOTONIC, &done);
+	not_done.tv_nsec += 100 * NSEC_PER_USEC;
+	done.tv_nsec += 300 * NSEC_PER_USEC;
```
@ClaCodes - it looks as though the 100 us and 300 us timeouts do not align with the 100 ms and 300 ms timeouts above.

```suggestion
	not_done.tv_nsec += 100 * NSEC_PER_MSEC;
	done.tv_nsec += 300 * NSEC_PER_MSEC;
```

cfriedt commented:
@ClaCodes - there is a subtle bug that needs to be fixed in the commit titled "posix: pthread: implement non-standard try-join and timed-join".

```diff
diff --git a/lib/posix/options/pthread.c b/lib/posix/options/pthread.c
index 5dfd6599784..db5d37fc8cc 100644
--- a/lib/posix/options/pthread.c
+++ b/lib/posix/options/pthread.c
@@ -1112,6 +1112,11 @@ static int pthread_timedjoin_internal(pthread_t pthread, void **status, k_timeou
        }
 
        ret = k_thread_join(&t->thread, timeout);
+       if (ret != 0) {
+               SYS_SEM_LOCK(&pthread_pool_lock) {
+                       t->attr.detachstate = PTHREAD_CREATE_JOINABLE;
+               }
+       }
        if (ret == -EBUSY) {
                return EBUSY;
        } else if (ret == -EAGAIN) {
```

ClaCodes commented:
> @ClaCodes - there is a subtle bug that needs to be fixed in the commit titled "posix: pthread: implement non-standard try-join and timed-join".
> 
> ```diff
> diff --git a/lib/posix/options/pthread.c b/lib/posix/options/pthread.c
> index 5dfd6599784..db5d37fc8cc 100644
> --- a/lib/posix/options/pthread.c
> +++ b/lib/posix/options/pthread.c
> @@ -1112,6 +1112,11 @@ static int pthread_timedjoin_internal(pthread_t pthread, void **status, k_timeou
>         }
>  
>         ret = k_thread_join(&t->thread, timeout);
> +       if (ret != 0) {
> +               SYS_SEM_LOCK(&pthread_pool_lock) {
> +                       t->attr.detachstate = PTHREAD_CREATE_JOINABLE;
> +               }
> +       }
>         if (ret == -EBUSY) {
>                 return EBUSY;
>         } else if (ret == -EAGAIN) {
> ```

@cfriedt  Thank you for sending this patch.
I have applied it and rebased.
Let me know, what you think

clamattia commented:
> > @ClaCodes - there is a subtle bug that needs to be fixed in the commit titled "posix: pthread: implement non-standard try-join and timed-join".
> > ```diff
> > diff --git a/lib/posix/options/pthread.c b/lib/posix/options/pthread.c
> > index 5dfd6599784..db5d37fc8cc 100644
> > --- a/lib/posix/options/pthread.c
> > +++ b/lib/posix/options/pthread.c
> > @@ -1112,6 +1112,11 @@ static int pthread_timedjoin_internal(pthread_t pthread, void **status, k_timeou
> >         }
> >  
> >         ret = k_thread_join(&t->thread, timeout);
> > +       if (ret != 0) {
> > +               SYS_SEM_LOCK(&pthread_pool_lock) {
> > +                       t->attr.detachstate = PTHREAD_CREATE_JOINABLE;
> > +               }
> > +       }
> >         if (ret == -EBUSY) {
> >                 return EBUSY;
> >         } else if (ret == -EAGAIN) {
> > ```
> 
> @cfriedt Thank you for sending this patch. I have applied it and rebased. Let me know, what you think

@cfriedt This seems to have fixed the tests. Could you resolve all conversations you consider resolved and then I will attempt to fix the rests (naming, return codes, ... etc.).

ycsin commented on lib/posix/options/pthread.c:
```diff
@@ -1115,8 +1111,18 @@ int pthread_join(pthread_t pthread, void **status)
+	ret = k_thread_join(&t->thread, timeout);
+	if (ret != 0) {
+		SYS_SEM_LOCK(&pthread_pool_lock) {
+			t->attr.detachstate = PTHREAD_CREATE_JOINABLE;
```
I don't get this, I think a comment here would be nice

ClaCodes replied:
Me neither. This the path @cfriedt send. Maybe, could you provide an explanatory comment?

cfriedt replied:
[Comment](https://github.com/zephyrproject-rtos/zephyr/blob/main/lib/posix/options/pthread.c#L1099) just a few lines above this in `main`.

cfriedt replied:
Only one thread can attempt to join another thread at a given time, so by changing the detached state to detached before calling `k_thread_join()` we can ensure that no other (p)threads could join it. However, if joining fails due to a timeout, we need to be able to retry later.

ClaCodes replied:
I have added a comment.

ycsin commented on tests/posix/common/src/pthread.c:
```diff
@@ -371,6 +377,67 @@ ZTEST(pthread, test_pthread_termination)
+	pthread_t th = {0};
+	void *retval;
+
+	/* Creating a threads that exits after 1s*/
```
The comments should match the updated sleep time

ClaCodes replied:
adjusted. Nice catch.

ycsin commented on tests/posix/common/src/pthread.c:
```diff
@@ -371,6 +377,67 @@ ZTEST(pthread, test_pthread_termination)
+	void *retval;
+
+	/* Creating a threads that exits after 1s*/
+	zassert_ok(pthread_create(&th, NULL, timedjoin_thread, NULL));
```
I think it would be nice to pass the sleep time as arg as well:

```suggestion
	useconds_t sleep_time;
	...
	/* Creating a threads that exits after 200ms */
	sleep_time = USEC_PER_MSEC * 200U;
	zassert_ok(pthread_create(&th, NULL, timedjoin_thread, (void *)sleep_time);
```

but probably just personal preference, so non-blocking

cfriedt replied:
It's not a bad suggestion, but `INT_TO_POINTER()` and `POINTER_TO_INT()` should be used.

cfriedt replied:
Also, we should use "create a thread" instead of "create a threads".

ClaCodes replied:
I have added this. But with type `int` instead of `useconds_t`, so that I could use the `INT_TO_POINTER`/`POINTER_TO_INT` macros. Let me know what you think.

ycsin commented on tests/posix/common/src/pthread.c:
```diff
@@ -371,6 +377,67 @@ ZTEST(pthread, test_pthread_termination)
+	zassert_equal(pthread_tryjoin_np(th, &retval), EBUSY);
+
+	/* Sleep so thread will exit */
+	usleep(USEC_PER_MSEC * 200U);
```
Could have use the `sleep_time` variable if my suggestion above is adopted

ClaCodes replied:
I generally like the sugestion. See my concern above.

ycsin commented on tests/posix/common/src/pthread.c:
```diff
@@ -371,6 +377,67 @@ ZTEST(pthread, test_pthread_termination)
+	zassert_ok(pthread_create(&th, NULL, timedjoin_thread, NULL));
+	usleep(USEC_PER_MSEC * 100U);
+
+	/* Attempting to join immediately should fail */
```
```suggestion
	/* Attempting to join without blocking should fail */
```
I was wondering why the comment says 'join immediately' when we slept for 100ms above

ClaCodes replied:
Fair enough. I adjusted to `/* Attempting to join, when thread is still running, should fail */`, what do you think?

ycsin commented on tests/posix/common/src/pthread.c:
```diff
@@ -371,6 +377,67 @@ ZTEST(pthread, test_pthread_termination)
+	/* Sleep so thread will exit */
+	usleep(USEC_PER_MSEC * 200U);
+
+	/* Attempting to join immediately should succeed now */
```
Just thought that this might be clearer (at least for me)
```suggestion
	/* Attempting to join without blocking should succeed now */
```

ClaCodes replied:
applied. Thanks.

cfriedt commented on tests/posix/common/src/pthread.c:
```diff
@@ -371,6 +377,67 @@ ZTEST(pthread, test_pthread_termination)
+	pthread_t th = {0};
+	void *retval;
+
+	/* Creating a threads that exits after 200ms*/
```
```suggestion
	/* Creating a thread that exits after 200ms*/
```

ClaCodes commented:
@cfriedt @ycsin Thanks for the suggestions and fixes.
I tried to apply them. Please have another look.

cfriedt commented on tests/posix/common/src/pthread.c:
```diff
@@ -143,6 +143,13 @@ static void *thread_top_exec(void *p1)
+static void *timedjoin_thread(void *p1)
+{
+	int sleep_duration_ms = POINTER_TO_INT(p1);
+	usleep(USEC_PER_MSEC * sleep_duration_ms);
```
@ClaCodes - this is causing the compliance failure.

```suggestion

	usleep(USEC_PER_MSEC * sleep_duration_ms);
```

ClaCodes replied:
Thanks, I added the newline.

ClaCodes commented:
@cfriedt @ycsin It looks like it is passing. Is there anything missing?

ClaCodes commented:
@ycsin @cfriedt Thank you for your support on this.
It still says 'Merging is blocked`. Do you require any more action from my side?

ycsin commented:
> @ycsin @cfriedt Thank you for your support on this.
> It still says 'Merging is blocked`. Do you require any more action from my side?

Just sit back and relax - you will be notified if there's another review/ when it get merged
